pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -416,166 +387,120 @@ public String segmentCommitEndWithMetadata(@QueryParam(SegmentCompletionProtocol
+      Preconditions.checkState(PinotSegmentUploadDownloadRestletResource.validateMultiPart(map, segmentName),
+          "Invalid multi-part for segment: %s", segmentName);
+      FormDataBodyPart bodyPart = map.values().iterator().next().get(0);
+
+      File localTempFile = new File(ControllerFilePathProvider.getInstance().getFileUploadTempDir(),
+          getTempSegmentFileName(segmentName));
+      try (InputStream inputStream = bodyPart.getValueAs(InputStream.class)) {
+        Files.copy(inputStream, localTempFile.toPath());
+      } catch (Exception e) {
+        FileUtils.deleteQuietly(localTempFile);
+        throw e;
       }
-      // Load segment metadata
-      return new SegmentMetadataImpl(tempMetadataDir);
-    } catch (Exception e) {
-      LOGGER.error("Exception extracting and reading segment metadata for {}", segmentNameStr, e);
-      return null;
+      return localTempFile;
     } finally {
-      FileUtils.deleteQuietly(tempMetadataDir);
+      form.cleanup();
     }
   }
 
   /**
-   *
-   * Extract a single file with name metaFileName from the input FormDataMultiPart and put it under the path
-   * tempMetadataDirStr + metaFileName.
-   * Return true iff the extraction and copy is successful.
+   * Extracts the segment metadata from the local segment file. Use the untarred file temporary directory to store the
+   * metadata files temporarily.
    */
-  private boolean extractMetadataFromInputField(FormDataMultiPart metadataFiles, String tempMetadataDirStr,
-      String metaFileName, String segmentName) {
-    FormDataBodyPart metadataFilesField = metadataFiles.getField(metaFileName);
-    Preconditions.checkNotNull(metadataFilesField, "The metadata input field %s does not exist.", metaFileName);
-
-    try (InputStream metadataPropertiesInputStream = metadataFilesField.getValueAs(InputStream.class)) {
-      Preconditions.checkNotNull(metadataPropertiesInputStream, "Unable to parse %s from input.", metaFileName);
-      java.nio.file.Path metadataPropertiesPath = FileSystems.getDefault().getPath(tempMetadataDirStr, metaFileName);
-      Files.copy(metadataPropertiesInputStream, metadataPropertiesPath);
-      return true;
-    } catch (IOException e) {
-      LOGGER.error("Failed to extract metadata property file: {} for segment {}", metaFileName, segmentName, e);
+  private static SegmentMetadataImpl extractMetadataFromLocalSegmentFile(File segmentFile)
+      throws Exception {
+    File tempIndexDir =
+        new File(ControllerFilePathProvider.getInstance().getUntarredFileTempDir(), segmentFile.getName());
+    try {
+      FileUtils.forceMkdir(tempIndexDir);
+
+      // Extract metadata.properties
+      try (InputStream metadataPropertiesInputStream = TarGzCompressionUtils
+          .unTarOneFile(new FileInputStream(segmentFile), V1Constants.MetadataKeys.METADATA_FILE_NAME)) {
+        Preconditions.checkState(metadataPropertiesInputStream != null, "Failed to find: %s from: %s",
```

Comment:
good to have the instance name that caused this request to fail or posted bad data

I will not repeat for other cases.

Reply:
Same here. The method only throws exception, and the caller will log all the extra information